include/sys/ddt_impl.h

```diff
@@ -51,6 +52,108 @@ extern "C" {
 	memcpy(&(ddlwe)->ddlwe_phys, (dde)->dde_phys, DDT_PHYS_SIZE(ddt)); \
 } while (0)
 
+#define	DDT_LOG_ENTRY_TO_LIGHTWEIGHT(ddt, ddle, ddlwe) do {             \
+	memset((ddlwe), 0, sizeof (*ddlwe));                            \
+	(ddlwe)->ddlwe_key = (ddle)->ddle_key;                          \
+	(ddlwe)->ddlwe_type = (ddle)->ddle_type;                        \
+	(ddlwe)->ddlwe_class = (ddle)->ddle_class;                      \
+	(ddlwe)->ddlwe_nphys = DDT_NPHYS(ddt);                          \
+	memcpy(&(ddlwe)->ddlwe_phys, (ddle)->ddle_phys, DDT_PHYS_SIZE(ddt)); \
+} while (0)
+
+/*
+ * An entry on the log tree. These are "frozen", and a record of what's in
+ * the on-disk log. They can't be used in place, but can be "loaded" back into
+ * the live tree.
+ */
+typedef struct {
+	ddt_key_t	ddle_key;	/* ddt_log_tree key */
+	avl_node_t	ddle_node;	/* ddt_log_tree node */
+
+	ddt_type_t	ddle_type;	/* storage type */
+	ddt_class_t	ddle_class;	/* storage class */
+
+	/* extra allocation for flat/trad phys */
+	ddt_univ_phys_t	ddle_phys[];
+} ddt_log_entry_t;
+
+/* On-disk log record types. */
+typedef enum {
+	DLR_INVALID	= 0,	/* end of block marker */
+	DLR_ENTRY	= 1,	/* an entry to add or replace in the log tree */
+} ddt_log_record_type_t;
+
+/* On-disk log record header. */
+typedef struct {
+	/*
+	 * dlr_info is a packed u64, use the DLR_GET/DLR_SET macros below to
+	 * access it.
+	 *
+	 * bits 0-7:	record type (ddt_log_record_type_t)
+	 * bits 8-23:	length of record header+payload
+	 * bits 24-31:	reserved, all zero
+	 * bits 32-39:	if type==DLR_ENTRY, storage type (ddt_type)
+	 *		  otherwise all zero
+	 * bits 40-47:	if type==DLR_ENTRY, storage class (ddt_class)
+	 *		  otherwise all zero
+	 * bits 48-63:	reserved, all zero
```

Comment:
I reordered it this way:

```c
+       /*
+        * dlr_info is a packed u64, use the DLR_GET/DLR_SET macros below to
+        * access it.
+        *
+        * bits 0-7:    record type (ddt_log_record_type_t)
+        * bits 8-15:   if type==DLR_ENTRY, storage type (ddt_type)
+        *                otherwise all zero
+        * bits 16-23:  if type==DLR_ENTRY, storage class (ddt_class)
+        *                otherwise all zero
+        * bits 24-39:  length of record header+payload
+        * bits 40-63:  reserved, all zero
+        */
```

I sorta like the idea of dividing down the length, but its a bit more cognitive overhead for unclear gain, as you say.

Reply:
Type and length are kind of universal header.  What would you say about this:
```
+       /*
+        * dlr_info is a packed u64, use the DLR_GET/DLR_SET macros below to
+        * access it.
+        *
+        * bits 0-7:    record type (ddt_log_record_type_t)
+        * bits 8-15:  length of record header+payload
+        * bits 16-47:  reserved, all zero
+        * bits 48-55:   if type==DLR_ENTRY, storage type (ddt_type)
+        *                otherwise all zero
+        * bits 56-63:  if type==DLR_ENTRY, storage class (ddt_class)
+        *                otherwise all zero
+        */
```

Reply:
Ahh yeah, I like that. Taken.